### ISSUE TITLE
Problems when downloading large files in small chunks

### ISSUE THREAD

**Issue by davidkazlauskas** [maidsafe/self_encryption#168]
I'm having problems downloading 5MB file when I request chunks anything less than that.

For instance, file is about 5MB, I request chunk 0-65536.

In the prepare_window function self.sequencer is expanded to accomodate 0-65536 elements.
(I'm talking about sources which are on crates.io at the moment, seems like things might have changed a little as I view source from this repository, if it was already fixed tell me)

However, downloaded chunk size seems exactly 1 MB. And it goes onto writing those chunks in sequencer which is currently only 0-65536 and "out of bounds" error happens. I've added a hack to extend sequencer one by one (probably won't be efficient with memory mapped files), it works now. Not sure if this is the best solution though. Left comments in all caps.

```
    /// Prepare a sliding window to ensure there are enough chunk slots for writing, and to read in
    /// any absent chunks from external storage.
    fn prepare_window(&mut self, length: u64, position: u64) {
        // EXPANSION - EXPANDS ONLY ACCORDING TO REQUESTED RANGE
        if (length + position) as usize > self.sequencer.len() {
            let tmp_size = self.sequencer.len();
            self.sequencer.extend(repeat(0).take((length as usize + position as usize) - tmp_size));
        }
        if self.file_size < (3 * MIN_CHUNK_SIZE) as u64 {
            return
        }
        let mut first_chunk = self.get_chunk_number(position);
        let mut last_chunk = self.get_chunk_number(position + length);
        if self.file_size < (3 * MAX_CHUNK_SIZE) as u64 {
            first_chunk = 0;
            last_chunk = 3;
        } else {
            for _ in 0..2 {
                if last_chunk < self.get_num_chunks() {
                    last_chunk += 1;
                }
            }
        }
        // [TODO]: Thread next - 2015-02-28 06:09pm
        let mut vec_deferred = Vec::new();
        for i in first_chunk..last_chunk {
            let mut found = false;
            for itr in self.chunks.iter() {
                if itr.number == i {
                    let pos = self.get_start_end_positions(i).0;
                    if itr.location == ChunkLocation::Remote {
                        vec_deferred.push(self.decrypt_chunk(i)
                            .chain::<_,String,_>(move |res|{
                                Ok((pos, res.unwrap()))
                            })
                        );
                    }
                    found = true;
                    break;
                }
            }
            if !found {
                self.chunks.push(Chunks{number: i, status: ChunkStatus::ToBeHashed,
                                        location: ChunkLocation::InSequencer});
            }
        }
        for (pos, vec) in
            Deferred::vec_to_promise(vec_deferred, ControlFlow::ParallelCPUS).sync().unwrap() {
            let mut pos_aux = pos;
            for itr2 in vec.iter() {
                // PREVIOUS VERSION - CRASHING
                //self.sequencer[pos_aux as usize] = *itr2;
                // HACK
                if self.sequencer.len() != (pos_aux as usize) {
                    self.sequencer[pos_aux as usize] = *itr2;
                } else {
                    self.sequencer.extend(repeat(*itr2).take(1));
                }
                pos_aux += 1;
            }
        }
    }

```

**Comment by lgvz** [comment]
This was fixed by commit 106de17.

**Comment by dirvine** [comment]
:+1 closing now.
